src/ctx.cpp

```diff
@@ -1630,7 +1635,7 @@ static int lArrayVectorWidth(llvm::Type *t) {
 }
 
 llvm::Value *FunctionEmitContext::BinaryOperator(llvm::Instruction::BinaryOps inst, llvm::Value *v0, llvm::Value *v1,
-                                                 const llvm::Twine &name) {
+                                                 bool isSigned, const llvm::Twine &name) {
```

Comment:
I was thinking about @aneshlya 's comment about unclear interface / design with passing this extra bool. This is very much true. The root of the problem that it should not be passed around, but rather all the decisions should be made locally based on the sign of the operation and/or operands. But this can be done only based on `ispc::Type`, which is not passed to `BinaryOperator()`. And changing this seems to be too intrusive change. An alternative is to make it more explicit than "some" bool value. I propose introducing a type for that and use it:

```c++
enum class WrapSemantics {
    NSW, // no need to preserve wraparound behavior in case of signed overflow
    None
};
```

Reply:
Done.

Reply:
`lEmitPrePostIncDec()`, `lEmitNegate()`, and `lEmitBinaryBitOp()` still use Boolean to carry wraparound semantics. Is this intentiaonal?

Reply:
Noted. For `lEmitBinaryOp`, the signedness information is used to determine `AShr` vs `LShr`, so it will remain for that function (although reverted to checking unsignedness to preserve behavior on booleans). Will pass through wrapSemantics for the others.